test/1.4/misc/T_dmllib.dml

```diff
@@ -0,0 +1,450 @@
+    TEST_ASSERT_MSG(_simple_event_data_representation(repr, indices, args)   \
+                    == _Simple_Event_Data_Format_ ## kind, __VA_ARGS__)
+
+    bool simple_event_data_representation_tests(char *msg_buf) {
+        attr_value_t indices = SIM_make_attr_list(
+            3,
+            SIM_make_attr_uint64(2),
+            SIM_make_attr_uint64(1),
+            SIM_make_attr_uint64(4));
+
+        attr_value_t args = SIM_make_attr_list(
+            3,
+            SIM_make_attr_boolean(true),
+            SIM_make_attr_int64(-4),
+            SIM_make_attr_floating(4.4));
+
+        attr_value_t indices_args = SIM_make_attr_list(2, indices, args);
+
+        attr_value_t entries[3];
+        for (int i = 0; i < 3; ++i) {
+            entries[i] = SIM_make_attr_list(2, SIM_make_attr_string("key"),
+                                            SIM_make_attr_nil());
+        }
+
+        attr_value_t one_entry_dict = SIM_make_attr_list(1, entries[0]);
+        attr_value_t two_entry_dict = SIM_make_attr_list(2, entries[1],
+                                                         entries[2]);
+
+        bool test_success = true;
+
+        // Tests for Legacy representation or invalid representations
+        TEST_REPR(indices, true, false, Legacy, "indices");
+        // The integer list might just as well be arguments
+        TEST_REPR(indices, false, true, Legacy, "indices/args");
+        TEST_REPR(indices, true, true, Invalid, "missing args");
+        TEST_REPR(indices, false, false, Invalid,
+                  "indices/args when not expected");
+
+
+        TEST_REPR(args, false, true, Legacy, "args");
+        TEST_REPR(args, true, false, Invalid, "args when not expected");
+        TEST_REPR(args, true, true, Invalid, "missing indices");
+
+        TEST_REPR(indices_args, true, true, Legacy, "indices & args");
+        TEST_REPR(indices_args, true, false, Invalid,
+                  "indices & args when only indices expected");
+        // [[index0, ...], [arg0, ..]] might just as well be as the serialized
+        // representation of two arguments
+        TEST_REPR(indices_args, false, true, Legacy,
+                  "indices & args when only args expected");
+        TEST_REPR(indices_args, false, false, Invalid,
+                  "indices & args when neither expected");
+
+        // Tests for pseudodict representation
+        // has_params, has_indices should be irrelevant
+        for (int i = 0; i < 4; ++i) {
+            bool has_indices = i & 1;
+            bool has_params  = i & 2;
+            TEST_REPR(one_entry_dict, has_indices, has_params, Pseudodict,
+                      "single-entry, indices=%d, params=%d",
+                      has_indices, has_params);
+            TEST_REPR(two_entry_dict, has_indices, has_params, Pseudodict,
+                      "double-entry, indices=%d, params=%d",
+                      has_indices, has_params);
+        }
+
+        exit:
```

Comment:
no need for a clean return on failure, fine to fail by FATAL_ERROR_IF or similar. Saves you the trouble of passing exit labels to macros, and propagating error message to caller. Also gives you the convenience of error messages with clickable line numbers that bring you to the failed test.

Reply:
Also note that with line numbers in error messages, the message is often redundant (exceptions include loop indices and when it replaces a comment).

Reply:
Now leveraging `ASSERT_MSG` and `ASSERT_FMT`.